Adds the ability to warn when unused parameters are set in a namespace.

evenator opened the pull request:
A common error when using unfamiliar ROS nodes is to accidentally set parameters by the wrong name. This feature allows the node author to output a WARNING for any unused parameters.

See the param_test node for an example.

evenator commented:
Yeah, I was following convention there. I can rewrite that as a unit test.

elliotjo commented on swri_roscpp/include/swri_roscpp/parameters.h:
```diff
@@ -1,15 +1,30 @@
+
+  /// This set stores all of the parameters that have been got with this library
+  /// It is used for the getUnusedParamKeys and WarnUnusedParams functions
+  std::set<std::string> _used_params;
```
This either needs to be declared static or extern.  extern makes more sense to work well with nodes that are spread across several files, but means swri_roscpp will have to produce an actual library that is linked into dependent code.  I'd also request that the prefix be changed from _ to g_ but that's a personal preference.

elliotjo commented on swri_roscpp/include/swri_roscpp/parameters.h:
```diff
@@ -126,8 +153,65 @@ namespace swri
+	return unused_params;
+  }
+
+  void warnUnusedParams(ros::NodeHandle const& n = ros::NodeHandle("~"))
```
Let's get rid of the default value.  It won't work with nodelets and we'll end up with lots of "we'll fix that later" problems.

evenator replied:
Sounds reasonable.

elliotjo commented on swri_roscpp/include/swri_roscpp/parameters.h:
```diff
@@ -126,8 +153,65 @@ namespace swri
+	std::vector<std::string>::iterator last_unused_param = std::set_difference(first_param_in_namespace, after_last_param_in_namespace, _used_params.begin(), _used_params.end(), unused_params.begin(), prefixLessThan);
+	unused_params.resize(last_unused_param - unused_params.begin());
+	return unused_params;
+  }
```
How about we go simpler here with something like:

std::vector<std::string> result;
for (blahblah it = all_params.begin(); it != all_params.end(); ++it) {
  if (*it.compare(0, n_namespace.size(), n) == 0 && used_params.count(*it) == 0) {
    result.push_back(*it);
  }
}

evenator replied:
I'll preface this by saying that I wrote this partly as an exercise in playing with the std algorithm library.

I agree that std algorithm code isn't easy on the eyes. However, the simpler algorithm would false-positive on members of arrays and structs. Admittedly, we don't yet have getters for those types, but that was the thinking. This filter is also faster than the simpler algorithm, which may or may not matter, depending on how many parameters are used by the node and are in the system.

Overall, I agree this needs a second pass to make it simpler, so I'll look into doing it.

evenator commented:
I've added an automated unit test for `getUnusedParameterKeys()`, per @pjreed's request. I've also removed the default value for the NodeHandle argument to `warnUnusedParameters()`, per @elliotjo's request.

Regarding simplifying the algorithm, I've concluded that the algorithm I wrote originally is correct, but I agree that it was inscrutable as written. I've attempted to alleviate that problem by adding more comments to describe what's going on. Is that an acceptable improvement?

evenator commented:
@pjreed Can you update your review? I added a unit test that can run in CI.

@pjreed, @elliotjo, @malban Is this merge-able now? Jason and Neal are interested in using it.

malban commented:
Fine with me
